doc/src/IOMMU.tex

```diff
@@ -843,6 +841,13 @@ \subsection{MSI PTE, MRIF mode}
 Any page address must be accepted for NPPN.
 \end{commentary}
 
+Memory accesses by I/O devices to addresses within a page
+covered by an MRIF-mode PTE are handled by the IOMMU
+instead of being passed through to the memory system.
+For a read access, the IOMMU returns zero as the read value.
+A write access is either interpreted as an MSI, resulting
+in an update of the destination MRIF, or is discarded.
```

Comment:
It is generally not natural for an IOMMU to provide data for a an IO read. Requiring an IOMMU to return 0 as the read value imposes a significant burden on the IOMMU construction to have a read data return path to the I/O bridge and for the I/O bridge to have a completion buffer to hold data that can be filled by either the IOMMU returned data or data returned from the memory subsystem. Requesting the IO bridge to complete with a value of 0 is also not natural for I/O bridges. 

Suggest that behavior of a read to a page covered by a MRIF-mode PTE be UNSPECIFIED. The IOMMU may either return zero as the read value or may request the IO bridge to abort the read as a Unsupported Request. 

A guideline to software to not program DMA to addresses within a page covered by an MRIF-mode PTE and to only program the address of the seteipnum_* register as appropriate into the registers used to program MSI addresses in the device.